### PR TITLE
Texture 3d

### PR DESCRIPTION
**Pull request by trungtle**
This sample demonstrates a use of 3d texture. Viewport is split into 4 to show rotation between each axis and with all three axes rotated

### CODE REVIEW

File: samples/texture_3d.html [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+        gl.bindBuffer(gl.ARRAY_BUFFER, null);
+
+        gl.bindVertexArray(null);
+
+        // -- Render
+        var orientation = [0.0, 0.0, 0.0];
+
+        requestAnimationFrame(render);
+
+        function yawPitchRoll(yaw, pitch, roll) {
```
Comment (pjcozzi) [inline samples/texture_3d.html:225]: If/when we use this in another sample, move it to utility.js.

Comment (pjcozzi) [inline samples/texture_3d.html:225]: Also, just FYI, in something like this, factor out all the redundant trig functions into locals for better performance, e.g.,

``` javascript
var cosPitch = Math.cos(pitch);
```

Maybe they would get optimized anyway, but I would not count on it.

File: samples/texture_3d.html [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+            ];
+        }
+
+        function render() {
+
+            orientation[0] += 0.020; // yaw
+            orientation[1] += 0.010; // pitch
+            orientation[2] += 0.005; // roll
+
+            var yawMatrix = new Float32Array(
```
Comment (pjcozzi) [inline samples/texture_3d.html]: Also, it is fine for this sample, but in a real app, we would avoid all this memory allocation per frame.

**Comment by pjcozzi** [comment]
Looks great, but I just want to make sure the swimming pattern is expected in the all the viewports except the top left.  Is that what the OpenGL sample looks like?

ok

**Comment by trungtle** [comment]
The swimming pattern was from using NEAREST filtering on the mipmap, instead of LINEAR filtering as in the opengl sample (which would get rid of the swimming pattern). The reason I switched was because I ran into an error in Firefox that said LINEAR_MIPMAP_LINEAR type undefined. Now, also the opengl sample uses a simplex noise generator which we don't have, but I don't know if that would make a different in term of showing the working of texture 3d.

**Comment by pjcozzi** [comment]
If `LINEAR_MIPMAP_LINEAR` works in Chrome, I would go with it.  Sounds like an issue with Firefox.

**Comment by pjcozzi** [comment]
It's not required, but if you want a noise function, I ported the Ashima Arts' GLSL version to JavaScript a few years ago, see noise3D.js: https://github.com/CIS565-Fall-2015/cis565-fall-2015.github.io/tree/master/webgl-examples/PerformanceTest

**Comment by trungtle** [comment]
Documented Firefox error here for reference: 

```
Error: WebGL: drawArraysInstanced: Active texture 0 for target 0x806f is 'incomplete', and will be rendered as RGBA(0,0,0,1), as per the GLES 2.0.24 $3.8.2: Because minification or magnification filtering is not NEAREST or NEAREST_MIPMAP_NEAREST, and the texture's format is a color format, its format must be "texture-filterable".
```

**Comment by trungtle** [comment]
@pjcozzi - when I'm added the noise3d.js and gl-matrix, gl-matrix doesn't seem to agree with Travis and it generates a lot of errors. Is it a good idea to keep it this way? I think it would be useful if we have gl-matrix eventually, but we'll have to modify Travis or gl-matrix to resolve these errors.

**Comment by pjcozzi** [comment]
Third-party libraries can be excluded from jsHint as I did in the `dev-shrek-texture-cube` branch (which was not merged):

https://github.com/WebGLSamples/WebGL2Samples/compare/dev-shrek-texture-cube#diff-b9e12334e9eafd8341a6107dd98510c9R7

Just make the same change.

This was discussed in https://github.com/WebGLSamples/WebGL2Samples/pull/42#issuecomment-169137231, please follow all the discussions in this repo.

**Comment by pjcozzi** [comment]
Also review the changes and discussion in #42 to see how to credit the third-party library.

**Comment by trungtle** [comment]
@pjcozzi this is good to go.

File: samples/noise3D.js [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+    function vec4_dot(left, right) {
+        return left[0] * right[0] +
+        left[1] * right[1] +
+        left[2] * right[2] +
+        left[3] * right[3];
+    }
+
+    //
+    // Description : Array and textureless GLSL 2D/3D/4D simplex
```
Comment (pjcozzi) [inline samples/noise3D.js]: Can you please move this to LICENSE.md (and format it similar to the other third-party libraries)?

**Comment by pjcozzi** [comment]
This is good.  Just please update the LICENSE.md and we'll merge.

File: samples/noise3D.js (hunk 3 above)
Comment (trungtle) [inline samples/noise3D.js]: Done!

**Comment by pjcozzi** [comment]
Looks good! :+1:

**Pull request merged**